vm/src/builtins/iter.rs

```diff
@@ -13,6 +14,91 @@ use crate::{
     TypeProtocol,
 };
 
+#[derive(Debug)]
+pub struct PositionIterInternal {
+    pub position: AtomicCell<usize>,
+    /// object or PyNone if exhausted
+    pub obj: PyRwLock<PyObjectRef>,
+}
+
+impl PositionIterInternal {
+    pub fn new(obj: PyObjectRef) -> Self {
+        Self {
+            position: AtomicCell::new(0),
+            obj: PyRwLock::new(obj),
+        }
+    }
+
+    pub fn is_active(&self, vm: &VirtualMachine) -> bool {
+        !vm.is_none(&self.obj.read())
+    }
+
+    pub fn set_state(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        if self.is_active(vm) {
+            if let Some(i) = state.payload::<PyInt>() {
+                let i = int::try_to_primitive(i.as_bigint(), vm).unwrap_or(0);
```

Comment:
Constrict this in `[0, len(obj)]` to follow with CPython. I don't think this leads to bad behavior in any of the other methods, it's only a visible difference when you directly call `__reduce__` from an iterator and see its output.

Reply:
I had add another function set_state_saturated() for the classes that should saturate the value.

Reply:
Ah I see. Isn't `set_state` _only_ used in the generic sequence iterator? Couldn't we just remove `set_state_saturated`, use the callable arg in `set_state` and pass `|_, pos| pos` for `PySequenceIterator` and `|obj, pos| pos.min(obj.len())` for the others?